### Olympic Medals with pandas!

In [1]:
import numpy as np
import pandas as pd

* web에서 csv를 불러오기.  
https://bit.ly/2YdeLFK

In [2]:
medals = pd.read_csv("https://bit.ly/2YdeLFK")

# display top 5 rows
display(medals.head())

# display last 5 rows
display(medals.tail())

City  Edition     Sport Discipline             Athlete  NOC Gender  \
0  Athens     1896  Aquatics   Swimming       HAJOS, Alfred  HUN    Men   
1  Athens     1896  Aquatics   Swimming    HERSCHMANN, Otto  AUT    Men   
2  Athens     1896  Aquatics   Swimming   DRIVAS, Dimitrios  GRE    Men   
3  Athens     1896  Aquatics   Swimming  MALOKINIS, Ioannis  GRE    Men   
4  Athens     1896  Aquatics   Swimming  CHASAPIS, Spiridon  GRE    Men   

                        Event Event_gender   Medal  
0              100m freestyle            M    Gold  
1              100m freestyle            M  Silver  
2  100m freestyle for sailors            M  Bronze  
3  100m freestyle for sailors            M    Gold  
4  100m freestyle for sailors            M  Silver

City  Edition      Sport       Discipline               Athlete  \
29211  Beijing     2008  Wrestling  Wrestling Gre-R        ENGLICH, Mirko   
29212  Beijing     2008  Wrestling  Wrestling Gre-R  MIZGAITIS, Mindaugas   
29213  Beijing     2008  Wrestling  Wrestling Gre-R       PATRIKEEV, Yuri   
29214  Beijing     2008  Wrestling  Wrestling Gre-R         LOPEZ, Mijain   
29215  Beijing     2008  Wrestling  Wrestling Gre-R        BAROEV, Khasan   

       NOC Gender       Event Event_gender   Medal  
29211  GER    Men   84 - 96kg            M  Silver  
29212  LTU    Men  96 - 120kg            M  Bronze  
29213  ARM    Men  96 - 120kg            M  Bronze  
29214  CUB    Men  96 - 120kg            M    Gold  
29215  RUS    Men  96 - 120kg            M  Silver

* medals에서 NOC를 추출하여 알파벳 순서대로 정렬한 뒤 countries에 저장하세요.

In [3]:
countries = np.sort(pd.unique(medals.NOC))
print(countries)

['AFG' 'AHO' 'ALG' 'ANZ' 'ARG' 'ARM' 'AUS' 'AUT' 'AZE' 'BAH' 'BAR' 'BDI'
 'BEL' 'BER' 'BLR' 'BOH' 'BRA' 'BUL' 'BWI' 'CAN' 'CHI' 'CHN' 'CIV' 'CMR'
 'COL' 'CRC' 'CRO' 'CUB' 'CZE' 'DEN' 'DJI' 'DOM' 'ECU' 'EGY' 'ERI' 'ESP'
 'EST' 'ETH' 'EUA' 'EUN' 'FIN' 'FRA' 'FRG' 'GBR' 'GDR' 'GEO' 'GER' 'GHA'
 'GRE' 'GUY' 'HAI' 'HKG' 'HUN' 'INA' 'IND' 'IOP' 'IRI' 'IRL' 'IRQ' 'ISL'
 'ISR' 'ISV' 'ITA' 'JAM' 'JPN' 'KAZ' 'KEN' 'KGZ' 'KOR' 'KSA' 'KUW' 'LAT'
 'LIB' 'LTU' 'LUX' 'MAR' 'MAS' 'MDA' 'MEX' 'MGL' 'MKD' 'MOZ' 'MRI' 'NAM'
 'NED' 'NGR' 'NIG' 'NOR' 'NZL' 'PAK' 'PAN' 'PAR' 'PER' 'PHI' 'POL' 'POR'
 'PRK' 'PUR' 'QAT' 'ROU' 'RSA' 'RU1' 'RUS' 'SCG' 'SEN' 'SIN' 'SLO' 'SRB'
 'SRI' 'SUD' 'SUI' 'SUR' 'SVK' 'SWE' 'SYR' 'TAN' 'TCH' 'TGA' 'THA' 'TJK'
 'TOG' 'TPE' 'TRI' 'TUN' 'TUR' 'UAE' 'UGA' 'UKR' 'URS' 'URU' 'USA' 'UZB'
 'VEN' 'VIE' 'YUG' 'ZAM' 'ZIM' 'ZZX']


* Iterating with a for loop
    - countries 내의 요소를 for loop로 하나씩 빼내서 print 하세요.

In [4]:
for country in countries:
    print(country)

AFG
AHO
ALG
ANZ
ARG
ARM
AUS
AUT
AZE
BAH
BAR
BDI
BEL
BER
BLR
BOH
BRA
BUL
BWI
CAN
CHI
CHN
CIV
CMR
COL
CRC
CRO
CUB
CZE
DEN
DJI
DOM
ECU
EGY
ERI
ESP
EST
ETH
EUA
EUN
FIN
FRA
FRG
GBR
GDR
GEO
GER
GHA
GRE
GUY
HAI
HKG
HUN
INA
IND
IOP
IRI
IRL
IRQ
ISL
ISR
ISV
ITA
JAM
JPN
KAZ
KEN
KGZ
KOR
KSA
KUW
LAT
LIB
LTU
LUX
MAR
MAS
MDA
MEX
MGL
MKD
MOZ
MRI
NAM
NED
NGR
NIG
NOR
NZL
PAK
PAN
PAR
PER
PHI
POL
POR
PRK
PUR
QAT
ROU
RSA
RU1
RUS
SCG
SEN
SIN
SLO
SRB
SRI
SUD
SUI
SUR
SVK
SWE
SYR
TAN
TCH
TGA
THA
TJK
TOG
TPE
TRI
TUN
TUR
UAE
UGA
UKR
URS
URU
USA
UZB
VEN
VIE
YUG
ZAM
ZIM
ZZX


* Make iterator country_it with countries
* Iterating over iterables : next()

In [5]:
country_it = iter(countries)

# print country_it with next()
print(next(country_it))

# two more times.
print(next(country_it))
print(next(country_it))

AFG
AHO
ALG


* Make dictionary olympic_dic with comprehesions.  
ex) {1896: 'Athens', 1900: 'Paris', ..., 2008: 'Beijing'}  
key : medals.Edition.   value : medals.City  
    - _HINT : using zip()!_

In [6]:
years = list(pd.unique(medals.Edition.values))
cities = []

for year in years:
    city_year = pd.unique(medals.loc[medals["Edition"]==year, 'City'])[0]
    cities.append(city_year)

print(cities)

# zip years and cities 
z = zip(years, cities)

# print list of the zip
z_list = list(z)
print(z_list)

['Athens', 'Paris', 'St Louis', 'London', 'Stockholm', 'Antwerp', 'Paris', 'Amsterdam', 'Los Angeles', 'Berlin', 'London', 'Helsinki', 'Melbourne / Stockholm', 'Rome', 'Tokyo', 'Mexico', 'Munich', 'Montreal', 'Moscow', 'Los Angeles', 'Seoul', 'Barcelona', 'Atlanta', 'Sydney', 'Athens', 'Beijing']
[(1896, 'Athens'), (1900, 'Paris'), (1904, 'St Louis'), (1908, 'London'), (1912, 'Stockholm'), (1920, 'Antwerp'), (1924, 'Paris'), (1928, 'Amsterdam'), (1932, 'Los Angeles'), (1936, 'Berlin'), (1948, 'London'), (1952, 'Helsinki'), (1956, 'Melbourne / Stockholm'), (1960, 'Rome'), (1964, 'Tokyo'), (1968, 'Mexico'), (1972, 'Munich'), (1976, 'Montreal'), (1980, 'Moscow'), (1984, 'Los Angeles'), (1988, 'Seoul'), (1992, 'Barcelona'), (1996, 'Atlanta'), (2000, 'Sydney'), (2004, 'Athens'), (2008, 'Beijing')]


In [7]:
olympic_dic = {year : city for year, city in z_list}
print(olympic_dic)

{1896: 'Athens', 1900: 'Paris', 1904: 'St Louis', 1908: 'London', 1912: 'Stockholm', 1920: 'Antwerp', 1924: 'Paris', 1928: 'Amsterdam', 1932: 'Los Angeles', 1936: 'Berlin', 1948: 'London', 1952: 'Helsinki', 1956: 'Melbourne / Stockholm', 1960: 'Rome', 1964: 'Tokyo', 1968: 'Mexico', 1972: 'Munich', 1976: 'Montreal', 1980: 'Moscow', 1984: 'Los Angeles', 1988: 'Seoul', 1992: 'Barcelona', 1996: 'Atlanta', 2000: 'Sydney', 2004: 'Athens', 2008: 'Beijing'}


* Iterating over dicionaries (olympic_dic)
    - olympic_dic의 key와 value를 하나씩 빼내서 print 하세요.

In [8]:
for key, value in olympic_dic.items():
    print(key, value)

1896 Athens
1900 Paris
1904 St Louis
1908 London
1912 Stockholm
1920 Antwerp
1924 Paris
1928 Amsterdam
1932 Los Angeles
1936 Berlin
1948 London
1952 Helsinki
1956 Melbourne / Stockholm
1960 Rome
1964 Tokyo
1968 Mexico
1972 Munich
1976 Montreal
1980 Moscow
1984 Los Angeles
1988 Seoul
1992 Barcelona
1996 Atlanta
2000 Sydney
2004 Athens
2008 Beijing


* Make dictionary olympic_dic1 with nested sturcture.  
    - 위에 했던 코드 응용하면 매우 좋습니다.  
    - winners는 해당 년도 올림픽의 전체 수상자 수입니다.  
ex) {1896: {'City': 'Athens', 'winners': 151}, 2008: {'City': 'Beijing', 'winners': 2042}}

In [9]:
winners = []
for index, year in enumerate(years):
    winners_year = len(medals[medals["Edition"]==year])
    winners.append(winners_year)
    
print(winners)

# zip cities, winners as z1, and zip years and z1
z1 = zip(cities, winners)
z2 = zip(years, z1)

# print list of the zip
z2_list = list(z2)
print(z2_list)

[151, 512, 470, 804, 885, 1298, 884, 710, 615, 875, 814, 889, 885, 882, 1010, 1031, 1185, 1305, 1387, 1459, 1546, 1705, 1859, 2015, 1998, 2042]
[(1896, ('Athens', 151)), (1900, ('Paris', 512)), (1904, ('St Louis', 470)), (1908, ('London', 804)), (1912, ('Stockholm', 885)), (1920, ('Antwerp', 1298)), (1924, ('Paris', 884)), (1928, ('Amsterdam', 710)), (1932, ('Los Angeles', 615)), (1936, ('Berlin', 875)), (1948, ('London', 814)), (1952, ('Helsinki', 889)), (1956, ('Melbourne / Stockholm', 885)), (1960, ('Rome', 882)), (1964, ('Tokyo', 1010)), (1968, ('Mexico', 1031)), (1972, ('Munich', 1185)), (1976, ('Montreal', 1305)), (1980, ('Moscow', 1387)), (1984, ('Los Angeles', 1459)), (1988, ('Seoul', 1546)), (1992, ('Barcelona', 1705)), (1996, ('Atlanta', 1859)), (2000, ('Sydney', 2015)), (2004, ('Athens', 1998)), (2008, ('Beijing', 2042))]


In [13]:
olympic_dic1 = {year : {'City' : subtup[0], 'winners' : subtup[1]} for year, subtup in z2_list}
print(olympic_dic1)

{1896: {'City': 'Athens', 'winners': 151}, 1900: {'City': 'Paris', 'winners': 512}, 1904: {'City': 'St Louis', 'winners': 470}, 1908: {'City': 'London', 'winners': 804}, 1912: {'City': 'Stockholm', 'winners': 885}, 1920: {'City': 'Antwerp', 'winners': 1298}, 1924: {'City': 'Paris', 'winners': 884}, 1928: {'City': 'Amsterdam', 'winners': 710}, 1932: {'City': 'Los Angeles', 'winners': 615}, 1936: {'City': 'Berlin', 'winners': 875}, 1948: {'City': 'London', 'winners': 814}, 1952: {'City': 'Helsinki', 'winners': 889}, 1956: {'City': 'Melbourne / Stockholm', 'winners': 885}, 1960: {'City': 'Rome', 'winners': 882}, 1964: {'City': 'Tokyo', 'winners': 1010}, 1968: {'City': 'Mexico', 'winners': 1031}, 1972: {'City': 'Munich', 'winners': 1185}, 1976: {'City': 'Montreal', 'winners': 1305}, 1980: {'City': 'Moscow', 'winners': 1387}, 1984: {'City': 'Los Angeles', 'winners': 1459}, 1988: {'City': 'Seoul', 'winners': 1546}, 1992: {'City': 'Barcelona', 'winners': 1705}, 1996: {'City': 'Atlanta', 'winn

* Iterating over dicionaries (olympic_dic1)
    - olympic_dic1의 key와 value를 하나씩 빼내서 print 하세요.

In [19]:
for key, value in olympic_dic1.items():
    print(key, value)

1896 {'City': 'Athens', 'winners': 151}
1900 {'City': 'Paris', 'winners': 512}
1904 {'City': 'St Louis', 'winners': 470}
1908 {'City': 'London', 'winners': 804}
1912 {'City': 'Stockholm', 'winners': 885}
1920 {'City': 'Antwerp', 'winners': 1298}
1924 {'City': 'Paris', 'winners': 884}
1928 {'City': 'Amsterdam', 'winners': 710}
1932 {'City': 'Los Angeles', 'winners': 615}
1936 {'City': 'Berlin', 'winners': 875}
1948 {'City': 'London', 'winners': 814}
1952 {'City': 'Helsinki', 'winners': 889}
1956 {'City': 'Melbourne / Stockholm', 'winners': 885}
1960 {'City': 'Rome', 'winners': 882}
1964 {'City': 'Tokyo', 'winners': 1010}
1968 {'City': 'Mexico', 'winners': 1031}
1972 {'City': 'Munich', 'winners': 1185}
1976 {'City': 'Montreal', 'winners': 1305}
1980 {'City': 'Moscow', 'winners': 1387}
1984 {'City': 'Los Angeles', 'winners': 1459}
1988 {'City': 'Seoul', 'winners': 1546}
1992 {'City': 'Barcelona', 'winners': 1705}
1996 {'City': 'Atlanta', 'winners': 1859}
2000 {'City': 'Sydney', 'winners':

* 아름다움! (문제에는 없었습니다.)

In [21]:
for key, value in olympic_dic1.items():
    print(key)
    for key1, value1 in value.items():
        print(key1, value1)

1896
City Athens
winners 151
1900
City Paris
winners 512
1904
City St Louis
winners 470
1908
City London
winners 804
1912
City Stockholm
winners 885
1920
City Antwerp
winners 1298
1924
City Paris
winners 884
1928
City Amsterdam
winners 710
1932
City Los Angeles
winners 615
1936
City Berlin
winners 875
1948
City London
winners 814
1952
City Helsinki
winners 889
1956
City Melbourne / Stockholm
winners 885
1960
City Rome
winners 882
1964
City Tokyo
winners 1010
1968
City Mexico
winners 1031
1972
City Munich
winners 1185
1976
City Montreal
winners 1305
1980
City Moscow
winners 1387
1984
City Los Angeles
winners 1459
1988
City Seoul
winners 1546
1992
City Barcelona
winners 1705
1996
City Atlanta
winners 1859
2000
City Sydney
winners 2015
2004
City Athens
winners 1998
2008
City Beijing
winners 2042


* By Using dictionary olympic_dic1
    - Sydney 올림픽은 몇 년에 열렸습니까?
    - 1972년에 열린 올림픽은 어느 도시에서 열렸습니까?
    - Beijing 올림픽의 전체 메달 수상자는 총 몇 명이었습니까?

In [11]:
# print first answer
for year in olympic_dic1.keys():
    if 'Sydney' in olympic_dic1.get(year).get('City'):
        answer = year
print("First answer is %s." %answer)

# print second answer
answer2 = olympic_dic1.get(1972).get('City')
print("Second answer is %s." %answer2)

# print third answer
for year in olympic_dic1.keys():
    if 'Beijing' in olympic_dic1.get(year).get('City'):
        temp = year
answer3 = olympic_dic1[temp]['winners']
print("Third answer is %s." %answer3)

First answer is 2000.
Second answer is Munich.
Third answer is 2042.
